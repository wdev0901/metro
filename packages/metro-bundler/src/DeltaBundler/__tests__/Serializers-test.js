/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails oncall+javascript_foundation
 * @format
 */

'use strict';

const Serializers = require('../Serializers');

const CURRENT_TIME = 1482363367000;

describe('Serializers', () => {
  const OriginalDate = global.Date;
  const getDelta = jest.fn();
  let deltaBundler;

  const deltaResponse = {
    pre: new Map([[1, {code: 'pre;'}]]),
    post: new Map([[2, {code: 'post;'}]]),
    delta: new Map([[3, {code: 'module3;'}], [4, {code: 'another;'}]]),
    inverseDependencies: [],
    reset: true,
  };

  function setCurrentTime(time: number) {
    global.Date = jest.fn(() => new OriginalDate(time));
  }

  beforeEach(() => {
    getDelta.mockReturnValueOnce(Promise.resolve(deltaResponse));

    deltaBundler = {
      async getDeltaTransformer() {
        return {
          id: '1234',
          deltaTransformer: {
            getDelta,
          },
        };
      },
    };

    setCurrentTime(CURRENT_TIME);
  });

  it('should return the stringified delta bundle', async () => {
    expect(
      await Serializers.deltaBundle(deltaBundler, {deltaBundleId: 10}),
    ).toMatchSnapshot();

    // Simulate a delta with some changes now
    getDelta.mockReturnValueOnce(
      Promise.resolve({
        delta: new Map([[3, {code: 'modified module;'}], [4, null]]),
        pre: new Map(),
        post: new Map(),
        inverseDependencies: [],
      }),
    );

    expect(
      await Serializers.deltaBundle(deltaBundler, {deltaBundleId: 10}),
    ).toMatchSnapshot();
  });

  it('should build the full JS bundle', async () => {
    expect(
      await Serializers.fullBundle(deltaBundler, {deltaBundleId: 10}),
    ).toMatchSnapshot();

    getDelta.mockReturnValueOnce(
      Promise.resolve({
        delta: new Map([[3, {code: 'modified module;'}], [4, null]]),
        pre: new Map([[5, {code: 'more pre;'}]]),
        post: new Map([[6, {code: 'bananas;'}], [7, {code: 'apples;'}]]),
        inverseDependencies: [],
      }),
    );
    setCurrentTime(CURRENT_TIME + 5000);

    expect(
      await Serializers.fullBundle(deltaBundler, {
        deltaBundleId: 10,
        sourceMapUrl: 'http://localhost:8081/myBundle.js',
      }),
    ).toMatchSnapshot();
  });

  // This test actually does not test the sourcemaps generation logic, which
  // is already tested in the source-map file.
  it('should build the full Source Maps', async () => {
    expect(
      await Serializers.fullSourceMap(deltaBundler, {deltaBundleId: 10}),
    ).toMatchSnapshot();

    getDelta.mockReturnValueOnce(
      Promise.resolve({
        delta: new Map([[3, {code: 'modified module;'}], [4, null]]),
        pre: new Map([[5, {code: 'more pre;'}]]),
        post: new Map([[6, {code: 'bananas;'}], [7, {code: 'apples;'}]]),
        inverseDependencies: [],
      }),
    );
    setCurrentTime(CURRENT_TIME + 5000);

    expect(
      await Serializers.fullSourceMap(deltaBundler, {deltaBundleId: 10}),
    ).toMatchSnapshot();
  });

  it('should return all the bundle modules', async () => {
    expect(
      await Serializers.getAllModules(deltaBundler, {deltaBundleId: 10}),
    ).toMatchSnapshot();

    getDelta.mockReturnValueOnce(
      Promise.resolve({
        delta: new Map([[3, {code: 'modified module;'}], [4, null]]),
        pre: new Map([[5, {code: 'more pre;'}]]),
        post: new Map([[6, {code: 'bananas;'}], [7, {code: 'apples;'}]]),
        inverseDependencies: [],
      }),
    );

    expect(
      await Serializers.getAllModules(deltaBundler, {deltaBundleId: 10}),
    ).toMatchSnapshot();
  });
});
