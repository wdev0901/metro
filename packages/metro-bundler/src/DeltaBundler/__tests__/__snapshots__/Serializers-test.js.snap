// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Serializers should build the full JS bundle 1`] = `
Object {
  "bundle": "pre;
module3;
another;
post;",
  "lastModified": 2016-12-21T23:36:07.000Z,
  "numModifiedFiles": 4,
}
`;

exports[`Serializers should build the full JS bundle 2`] = `
Object {
  "bundle": "pre;
more pre;
modified module;
post;
bananas;
apples;",
  "lastModified": 2016-12-21T23:36:12.000Z,
  "numModifiedFiles": 5,
}
`;

exports[`Serializers should build the full Source Maps 1`] = `"{\\"version\\":3,\\"sources\\":[],\\"sourcesContent\\":[],\\"names\\":[],\\"mappings\\":\\"\\"}"`;

exports[`Serializers should build the full Source Maps 2`] = `"{\\"version\\":3,\\"sources\\":[],\\"sourcesContent\\":[],\\"names\\":[],\\"mappings\\":\\"\\"}"`;

exports[`Serializers should return all the bundle modules 1`] = `
Array [
  Object {
    "code": "pre;",
  },
  Object {
    "code": "module3;",
  },
  Object {
    "code": "another;",
  },
  Object {
    "code": "post;",
  },
]
`;

exports[`Serializers should return all the bundle modules 2`] = `
Array [
  Object {
    "code": "pre;",
  },
  Object {
    "code": "more pre;",
  },
  Object {
    "code": "modified module;",
  },
  Object {
    "code": "post;",
  },
  Object {
    "code": "bananas;",
  },
  Object {
    "code": "apples;",
  },
]
`;

exports[`Serializers should return the stringified delta bundle 1`] = `
Object {
  "bundle": "{\\"id\\":\\"1234\\",\\"pre\\":[[1,\\"pre;\\"]],\\"post\\":[[2,\\"post;\\"]],\\"delta\\":[[3,\\"module3;\\"],[4,\\"another;\\"]],\\"reset\\":true}",
  "numModifiedFiles": 4,
}
`;

exports[`Serializers should return the stringified delta bundle 2`] = `
Object {
  "bundle": "{\\"id\\":\\"1234\\",\\"pre\\":[],\\"post\\":[],\\"delta\\":[[3,\\"modified module;\\"],[4,null]]}",
  "numModifiedFiles": 2,
}
`;
